dds/idl/marshal_generator.cpp

```diff
@@ -1873,66 +1900,65 @@ bool marshal_generator::gen_struct(AST_Structure* node,
     extraction.addArg("stru", cxx + "&");
     extraction.endArgs();
     string intro;
-    if (maybe_delimited) {
+    if (may_be_delimited) {
       be_global->impl_ <<
-        "  unsigned total_size;\n";
+        "  size_t total_size = 0;\n";
       const char* indent = "  ";
       if (not_only_delimited) {
         indent = "    ";
         be_global->impl_ <<
-          "  if (strm.xcdr_version() == Serializer::XCDR2) {\n";
+          "  if (strm.encoding().xcdr_version() == Encoding::XCDR_VERSION_2) {\n";
       }
       be_global->impl_ <<
-        indent << "if (!strm.read_delimiter(size)) {\n" <<
+        indent << "if (!strm.read_delimiter(total_size)) {\n" <<
         indent << "  return false;\n" <<
         indent << "}\n";
       if (not_only_delimited) {
         be_global->impl_ <<
           "  }\n";
       }
     }
-    if (parameter_list) {
+    if (may_be_parameter_list) {
       if (repr.xcdr2) {
         /**
-         * We don't have a sentinel pid in XCDR2 paramter lists, but we have
+         * We don't have a sentinel pid in XCDR2 parameter lists, but we have
          * the size, so we need to stop after we hit this offset.
          */
         be_global->impl_ <<
-          "  const size_t end_of_fields = pos_rd() + total_size;\n";
+          "  const char* end_of_fields = strm.pos_rd() + total_size;\n";
       }
       be_global->impl_ <<
-        "  unsigned field_id;\n"
+        "  unsigned member_id;\n"
         "  size_t field_size;\n"
-        "  while (true) {\n"
-        "    if (!strm.read_parameter_id(field_id, field_size)) {\n"
-        "      return false;\n"
-        "    }\n";
-      if (repr.xcdr1) {
-        be_global->impl_ <<
-          "    if (member_id == Serializer::pid_list_end";
-        if (repr.not_only_xcdr1()) {
+        "  while (true) {\n";
+
+      if (repr.xcdr2) {
+        be_global->impl_ << "\n"
+          "    if ((!strm.pos_rd() || strm.pos_rd() >= end_of_fields)";
```

Comment:
Comparing values of addresses in `strm.pos_rd()` and `end_of_fields`? I'm not sure if this is portable.

Reply:
That's a good question. It's probably not, but it needs to be done somehow. I think it's safe for us to assume we can cast a pointer to an integer that we can compare, but I think it will need more work as appendable is implemented.

Reply:
Can we use offsets in bytes to do the comparison?